source/grid/grid_generator_pipe_junction.cc

```diff
@@ -0,0 +1,436 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/manifold.h>
+
+#include <deal.II/physics/transformations.h>
+#include <deal.II/physics/vector_relations.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace
+{
+  /**
+   * Auxiliary constructs for the pipe junction geometry.
+   *
+   * Please refer to the in-source documentation of the pipe_junction function
+   * below for more information about the individual contents.
+   */
+  namespace PipeSegment
+  {
+    /**
+     * Selection of pipe segment properties to calculate its height with the
+     * function below.
+     */
+    struct AdditionalData
+    {
+      double skeleton_length;
+
+      double cosecant_polar;
+      double cotangent_polar;
+      double cotangent_azimuth_half_right;
+      double cotangent_azimuth_half_left;
+    };
+
+
+
+    /**
+     * Calculate the height of a pipe segment, depending on the location in the
+     * x-y plane.
+     */
+    inline double
+    compute_z_expansion(const double          x,
+                        const double          y,
+                        const AdditionalData &data)
+    {
+      return
+        // Scale the unit cylinder to the correct length.
+        data.skeleton_length
+        // Next, adjust for the polar angle. This part will be zero if all
+        // openings and the bifurcation are located on a plane.
+        + x * data.cotangent_polar
+        // Last, adjust for the azimuth angle.
+        - std::abs(y) * data.cosecant_polar *
+            ((y > 0) ? data.cotangent_azimuth_half_right :
+                       data.cotangent_azimuth_half_left);
+    }
+  } // namespace PipeSegment
+} // namespace
+
+
+
+namespace GridGenerator
+{
+  template <int dim, int spacedim>
+  void
+  pipe_junction(Triangulation<dim, spacedim> &,
+                const std::vector<std::pair<Point<spacedim>, double>> &,
+                const std::pair<Point<spacedim>, double> &,
+                const double)
+  {
+    Assert(false, ExcNotImplemented());
+  }
+
+
+
+  template <>
+  void
+  pipe_junction(Triangulation<3, 3> &                           tria,
+                const std::vector<std::pair<Point<3>, double>> &openings,
```

Comment:
```suggestion
                const std::array<std::pair<Point<3>, double>, 3> &openings,
```
This would be more explicit IMO.

Reply:
We should also be explicit in the documentation that `openings` has to be of size 3.

Reply:
I'd also rather have an std::array, but we could fix this later when we fix `Patterns::Tools::Convert`.

Reply:
Okay, I will open a separate issue for this, and give some supplementary information.